cluster/ci/origin/logging-infra/gcs-sidecar.go

```diff
@@ -0,0 +1,74 @@
+// gcs-sidecar shares a volume with entrypoint exposes that process
+// to the larger CI infrastructure. At startup, metadata is recorded
+// and uploaded to GCS. During execution, the process log is streamed
+// to GCS. When the marker file is placed, exit metadata are recorded
+// and uploaded alongside artifacts to GCS.
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"io/ioutil"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/storage"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		log.Fatalf("USAGE: %s CONFIGURATION", os.Args[0])
+	}
+
+	config, err := loadConfig(os.Args[1])
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	gcsBucket, err := createGcsClient(config.GcsBucket)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := UploadStartingData(config.MetadataFile, gcsBucket); err != nil {
+		log.Fatal(err)
+	}
+
+	returnCode, err := retrieveProcessReturnCode(config.MarkerFile)
+	if err != nil {
+		// even if we fail to determine this info,
+		// we want to upload as much as we can
+		log.Printf("failed to determine process return code: %v", err)
+	}
+	passed := returnCode == 0 && err == nil
+
+	if err := UploadFinishedData(config.ProcessLog, config.MetadataFile, config.ArtifactDir, passed, gcsBucket); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func createGcsClient(bucket string) (*storage.BucketHandle, error) {
+	ctx := context.Background()
+	gcsClient, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to GCS: %v", err)
+	}
+	return gcsClient.Bucket(bucket), nil
+}
+
+// retrieveProcessReturnCode waits until the return code
+// is available in the marker file, then returns it
+func retrieveProcessReturnCode(markerFile string) (int, error) {
+	for {
+		data, err := ioutil.ReadFile(markerFile)
```

Comment:
may be overkill, but there is https://github.com/fsnotify/fsnotify

in some dream world of the future I expect runs to be fast... potentially adding 10 secs to the end of it may be too long :)